state/protocol/badger/mutator.go

```diff
@@ -805,115 +792,74 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
-	if err != nil {
-		if errors.Is(err, storage.ErrNotFound) {
-			return nil, nil, fmt.Errorf("ordering seals: parent payload contains seals for unknown block: %s", err.Error())
-		}
-		return nil, nil, fmt.Errorf("unexpected error ordering seals: %w", err)
-	}
-
-	// track service event driven metrics and protocol events that should be emitted
-	for _, seal := range orderedSeals {
-		result, err := m.results.ByID(seal.ResultID)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not retrieve result (id=%x) for seal (id=%x): %w", seal.ResultID, seal.ID(), err)
-		}
-		for _, event := range result.ServiceEvents {
-			switch ev := event.Event.(type) {
-			case *flow.EpochSetup:
-				// update current epoch phase
-				events = append(events, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseSetup) })
-				// track epoch phase transition (staking->setup)
-				events = append(events, func() { m.consumer.EpochSetupPhaseStarted(ev.Counter-1, block.Header) })
-			case *flow.EpochCommit:
-				// update current epoch phase
-				events = append(events, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseCommitted) })
-				// track epoch phase transition (setup->committed)
-				events = append(events, func() { m.consumer.EpochCommittedPhaseStarted(ev.Counter-1, block.Header) })
-			default:
-				// do nothing
-			}
-		}
-	}
-
-	return
+	if finalizedPsSnapshot.InvalidEpochTransitionAttempted() {
+		// No epoch state changes to notify on when EFM is triggered
+		return nil, nil, nil
+	}
+
+	parentEpochCounter := parentPsSnapshot.Epoch()
+	childEpochCounter := finalizedPsSnapshot.Epoch()
+	parentEpochPhase := parentPsSnapshot.EpochPhase()
+	childEpochPhase := finalizedPsSnapshot.EpochPhase()
+
+	// Different counter -> must be an epoch transition
+	if parentEpochCounter != childEpochCounter {
+		childEpochSetup := finalizedPsSnapshot.EpochSetup()
+		events = append(events, func() { m.consumer.EpochTransition(childEpochSetup.Counter, finalized) })
+		// set current epoch counter corresponding to new epoch
+		metrics = append(metrics, func() { m.metrics.CurrentEpochCounter(childEpochSetup.Counter) })
+		// denote the most recent epoch transition height
+		metrics = append(metrics, func() { m.metrics.EpochTransitionHeight(finalized.Height) })
+		// set epoch phase - since we are starting a new epoch we begin in the staking phase
+		metrics = append(metrics, func() { m.metrics.CurrentEpochPhase(flow.EpochPhaseStaking) })
+		// set current epoch view values
+		metrics = append(
+			metrics,
+			func() { m.metrics.CurrentEpochFinalView(childEpochSetup.FinalView) },
+			func() { m.metrics.CurrentDKGPhase1FinalView(childEpochSetup.DKGPhase1FinalView) },
+			func() { m.metrics.CurrentDKGPhase2FinalView(childEpochSetup.DKGPhase2FinalView) },
+			func() { m.metrics.CurrentDKGPhase3FinalView(childEpochSetup.DKGPhase3FinalView) },
+		)
+		return
+	}
+	// Same counter and same phase -> nothing to do
+	if parentEpochPhase == childEpochPhase {
```

Comment:
I don't a situation where we will have same phase but different counters, maybe we can move this check before the epoch counter check? This will explicitly mean that we report metrics only on phase changes.

Reply:
👍 updated in https://github.com/onflow/flow-go/pull/5923/commits/c115206330a0efec8da8cda30acc770fc9f2c6bc